java/java.lsp.server/vscode/src/propertiesView/controlTypes.ts

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright (c) 2023, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl.
+ */
+import { EnumType, KeyOfArray, Typed } from "../typesUtil";
+
+export type ID = number;
+
+export const PropTypes: KeyOfArray<PropTypeMap> = [
+    "java.lang.String",
+    "java.lang.Boolean",
+    "java.util.Properties",
+    "unknown"
+];// unfortunate but necessary duplication
+export type PropTypeMap = {
+    "java.lang.String": string;
+    "java.lang.Boolean": boolean;
+    "java.util.Properties": Record<string, string>;
+    "unknown": unknown
+};
+export type Property<T extends keyof PropTypeMap = keyof PropTypeMap> = T extends T ? {
+    propPref: boolean;
+    propDispName: string;
+    propShortName: string;
+    propWrite: boolean;
+    propHidden: boolean;
+    propExpert: boolean;
+    propType: T;
+    propValue: PropTypeMap[T];
+    propName: string;
+} : never; // Distributive type
+export type Properties = {
+    propPref: boolean;
+    propDispName: string;
+    propShortName: string;
+    propHidden: boolean;
+    propExpert: boolean;
+    propName: string;
+    props: Property[];
+};
+
+export type MessageProp = {
+    name: string;
+    value: string | boolean | Record<string, string>;
+};
+
+export type Command = "Save" | "Cancel" | "Error" | "Info";
```

Comment:
I've just tried a different approach: declare a string-valued enum-like with `Save`, 'Cancel` ... members, make an union `CommandType` type:
```
export const CommandKey = {
    Save : "Save",
    Error : "Error",
    Info : "Info",
    Cancel : "Cancel"
} as const;

export type CommandType = (typeof CommandKey)[keyof typeof CommandKey];

export type MessageCommon<T extends CommandType> = Typed<T>;
```
(rest of code unchanged) ... but I am not entirely sure about benefits of such approach, it does not save much.
But I'd at least suggest to make the union type should use an `CommandKey` enum's values, to save string duplication:

```
export enum CommandKey {
    Save = "Save",
    Error = "Error",
    Info = "Info",
    Cancel = "Cancel"
}

export type Command = CommandKey.Save | CommandKey.Cancel | CommandKey.Error | CommandKey.Info;
```

Reply:
The code is made so it is imperative to be complete...
The Value of 'enum' has to be exactly the same as the Key and has to contain all of Type union and nothing more...
Found no other way of achieving this behavior...